dask/dataframe/tests/test_io.py

```diff
@@ -758,6 +783,169 @@ def test_to_hdf_multiple_datasets():
+        out = dd.read_hdf(fn, '/data')
+        eq(df.append(df), out)
+
+    # overwriting a file with two partitions
+    a = dd.from_pandas(df, 2)
+    with tmpdir() as dn:
+        fn = os.path.join(dn, 'data*')
+        a.to_hdf(os.path.join(dn, 'data1'), '/data')
+        a.to_hdf(fn, '/data', mode='w')
+        out = dd.read_hdf(fn, '/data')
+        eq(df, out)
+
+    # overwriting a single partition, keeping other partitions
+    a = dd.from_pandas(df, 2)
+    with tmpdir() as dn:
+        fn = os.path.join(dn, 'data*')
+        a.to_hdf(os.path.join(dn, 'data1'), '/data')
+        a.to_hdf(fn, '/data', mode='a', append=False)
+        out = dd.read_hdf(fn, '/data')
+        eq(df.append(df), out)
+
+
+def test_to_hdf_link_optimizations():
+    """testing dask link levels is correct by calculating the depth of the dask graph"""
+    pytest.importorskip('tables')
+    df16 = pd.DataFrame({'x': ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p'],
+                       'y': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]},
+                            index=[1., 2., 3., 4., 5., 6., 7., 8., 9., 10., 11., 12., 13., 14., 15., 16.])
+    a = dd.from_pandas(df16, 16)
+
+    # saving to multiple hdf files, no links are needed
+    # expected layers: from_pandas, to_hdf, list = depth of 3
+    with tmpdir() as dn:
+        fn = os.path.join(dn, 'data*')
+        d = a.to_hdf(fn, '/data', compute=False)
+        assert dependency_depth(d.dask) == 3
+
+    # saving to a single hdf file with multiple nodes
+    # all subsequent nodes depend on the first
+    # expected layers: from_pandas, first to_hdf(creates file+node), subsequent to_hdfs, list = 4
+    with tmpfile() as fn:
+        d = a.to_hdf(fn, '/data*', compute=False)
+        assert dependency_depth(d.dask) == 4
+
+    # saving to a single hdf file with a single node
+    # every node depends on the previous node
+    # expected layers: from_pandas, to_hdf times npartitions(15), list = 2 + npartitions = 17
+    with tmpfile() as fn:
+        d = a.to_hdf(fn, '/data', compute=False)
+        assert dependency_depth(d.dask) == 2 + a.npartitions
+
+
+@pytest.mark.slow
+def test_to_hdf_lock_delays():
+    pytest.importorskip('tables')
+    df16 = pd.DataFrame({'x': ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p'],
+                       'y': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]},
+                            index=[1., 2., 3., 4., 5., 6., 7., 8., 9., 10., 11., 12., 13., 14., 15., 16.])
+    a = dd.from_pandas(df16, 16)
+
+    def delayed_nop(i):
+        sleep(0.1*(16-i[1]))
+        return i
+
+    # saving to multiple hdf nodes
+    # adding artifichial delays to make sure last tasks finish first
+    with tmpfile() as fn:
+        a = a.apply(delayed_nop, axis=1, columns=a.columns)
+        a.to_hdf(fn, '/data*')
+        out = dd.read_hdf(fn, '/data*')
+        eq(df16, out)
```

Comment:
It was a simple way to make sure results don't arrive in their expected order. I can speed it up by only delaying some of them.